src/Filter.ts

```diff
@@ -0,0 +1,110 @@
+export interface FilterArgs {
+	[ name: string ]: Filter | RegExp | string | (() => boolean);
+}
+
+export default class Filter {
+	[index: string]: any;
+	type: string;
+	args: any[];
+
+	constructor();
+	constructor(filterArg: string);
+	constructor(filterArg: () => any);
+	constructor(filterArg: FilterArgs);
+	constructor(filterArg?: any) {
+		if (typeof filterArg === 'string' || typeof filterArg === 'function') {
+			this.type = typeof filterArg;
+			this.args = [filterArg];
+		}
+		else if (filterArg) {
+			var filter = this;
+			// construct a filter based on the query object
+			for (var key in filterArg) {
+				var value = filterArg[key];
+				if (value instanceof Filter) {
+					// fully construct the filter from the single arg
+					filter = filter[value.type](key, value.args[0]);
+				}
+				else if ((<any>value) && typeof (<any>value).test === 'function') {
```

Comment:
Do you need the first cast: `(<any> value)`? Couldn't it just be `if (value && (<any> value).test === 'function')`

Also, I don't think this is in the style guide, but it is in our tslint.json for most of the dojo2 projects to put a space between the cast and the variable: `(<any> value)` as opposed to `(<any>value)`; it's a small detail but I just thought I would mention it.